### PR TITLE
Feature/fix backlink

### PR DESCRIPTION
**Pull request by graham-willis-druid**
Added smarter backlink functionality

### CODE REVIEW

File: packages/gafl-webapp-service/src/pages/summary/contact-summary/__tests__/contact-summary.spec.js [hunk 1]
```diff
@@ -99,6 +80,8 @@ describe('The contact summary page', () => {
     const data = await injectWithCookies('GET', CONTACT_SUMMARY.uri)
     expect(data.statusCode).toBe(302)
     expect(data.headers.location).toBe(NAME.uri)
+    const data2 = await injectWithCookies('GET', NAME.uri)
+    expect(data2.payload.search(backLinkRegEx(LICENCE_SUMMARY.uri)) > 0).toBeTruthy()
```
Comment (jaucourt) [inline packages/gafl-webapp-service/src/pages/summary/contact-summary/__tests__/contact-summary.spec.js:84]: This looks like it should be a separate, new test

File: packages/gafl-webapp-service/src/pages/summary/contact-summary/__tests__/contact-summary.spec.js [hunk 2]
```diff
@@ -114,9 +97,9 @@ describe('The contact summary page', () => {
 
   it('redirects to the contact page if no contact details have been set', async () => {
     await postRedirectGet(ADDRESS_ENTRY.uri, goodAddress)
-    const data = await injectWithCookies('GET', CONTACT_SUMMARY.uri)
-    expect(data.statusCode).toBe(302)
-    expect(data.headers.location).toBe(CONTACT.uri)
+    const data2 = await injectWithCookies('GET', CONTACT_SUMMARY.uri)
```
Comment (jaucourt) [inline packages/gafl-webapp-service/src/pages/summary/contact-summary/__tests__/contact-summary.spec.js:100]: Why did you rename this variable? `data` is probably clearer, `data2` makes you wonder what happened to `data`

File: packages/gafl-webapp-service/src/pages/summary/contact-summary/__tests__/contact-summary.spec.js [hunk 3]
```diff
@@ -135,108 +118,53 @@ describe('The contact summary page', () => {
...
-    const data = await injectWithCookies('GET', CONTROLLER.uri)
-    expect(data.statusCode).toBe(302)
-    expect(data.headers.location).toBe(LICENCE_SUMMARY.uri)
-    await injectWithCookies('GET', LICENCE_SUMMARY.uri)
-    await injectWithCookies('POST', LICENCE_SUMMARY.uri)
-
-    const data2 = await injectWithCookies('GET', CONTROLLER.uri)
+    const data = await injectWithCookies('GET', ADDRESS_ENTRY.uri)
+    expect(data.payload.search(backLinkRegEx(CONTACT_SUMMARY.uri)) > 0).toBeTruthy()
+    const data2 = await postRedirectGet(ADDRESS_ENTRY.uri, goodAddress)
```
Comment (jaucourt) [inline packages/gafl-webapp-service/src/pages/summary/contact-summary/__tests__/contact-summary.spec.js:148]: Have all the redirects on the left stopped happening?

File: packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js [hunk 4]
```diff
@@ -107,33 +117,41 @@ describe('The licence summary page', () => {
   })
 
   it('licence length amendment causes a redirect to the summary page', async () => {
+    await injectWithCookies('GET', LICENCE_SUMMARY.uri)
+    const data = await injectWithCookies('GET', LICENCE_LENGTH.uri)
+    // Tests that the back-link is now set to the summary page
+    expect(data.payload.search(backLinkRegEx(LICENCE_SUMMARY.uri)) > 0).toBeTruthy()
```
Comment (jaucourt) [inline packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js:123]: Should this be a separate test (that the back link updates), to keep tests simple?

File: packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js [hunk 5]
```diff
@@ -107,33 +117,41 @@ describe('The licence summary page', () => {
...
+    expect(data2.statusCode).toBe(200)
+    const data3 = await postRedirectGet(LICENCE_TYPE.uri, { 'licence-type': 'trout-and-coarse' })
     expect(data3.statusCode).toBe(302)
-    expect(data3.headers.location).toBe(LICENCE_SUMMARY.uri)
+    expect(data3.headers.location).toBe(NUMBER_OF_RODS.uri)
+    const data4 = await injectWithCookies('GET', NUMBER_OF_RODS.uri)
+    expect(data4.payload.search(backLinkRegEx(LICENCE_SUMMARY.uri)) > 0).toBeTruthy()
+    const data5 = await postRedirectGet(NUMBER_OF_RODS.uri, { 'number-of-rods': '2' })
+    expect(data5.statusCode).toBe(302)
+    expect(data5.headers.location).toBe(LICENCE_SUMMARY.uri)
```
Comment (jaucourt) [inline packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js:144]: This test is starting to feel really complex - I can see that you're trying to test an eventual redirect, but there's quite a lot of other stuff being tested too, might be better to break these out to other tests?

File: packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js [hunk 6]
```diff
@@ -220,4 +238,70 @@ describe('The licence summary page', () => {
...
+    expect(data2.headers.location).toBe(NAME.uri)
+  })
+
+  it('date of birth amendment (junior) causes a method of contact of letter to be set no none', async () => {
+    await injectWithCookies('POST', DATE_OF_BIRTH.uri, dobHelper(dobSeniorToday))
+    await injectWithCookies('GET', CONTROLLER.uri)
+    await postRedirectGet(CONTACT.uri, { 'how-contacted': 'none' })
+    const { payload } = await injectWithCookies('GET', TEST_TRANSACTION.uri)
+    expect(JSON.parse(payload).permissions[0].licensee.preferredMethodOfConfirmation).toEqual(HOW_CONTACTED.letter)
+    expect(JSON.parse(payload).permissions[0].licensee.preferredMethodOfReminder).toEqual(HOW_CONTACTED.letter)
```
Comment (jaucourt) [inline packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js:299]: This seems to diverge from the stated ambition of the test - we're testing for a letter contact? I think this should be a separate test

File: packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js [hunk 7]
```diff
@@ -220,4 +238,70 @@ describe('The licence summary page', () => {
...
+  })
+
+  it('date of birth amendment (junior) causes redirect to the junior licence page and subsequent redirect to the licence summary page and then the contact summary page', async () => {
+    await injectWithCookies('POST', DATE_OF_BIRTH.uri, dobHelper(dobJuniorToday))
+    let data = await injectWithCookies('GET', CONTROLLER.uri)
+    expect(data.statusCode).toBe(302)
+    expect(data.headers.location).toBe(JUNIOR_LICENCE.uri)
+    data = await postRedirectGet(JUNIOR_LICENCE.uri, {})
+    expect(data.statusCode).toBe(302)
+    expect(data.headers.location).toBe(LICENCE_SUMMARY.uri)
```
Comment (jaucourt) [inline packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js:283]: Could the further assertions after these be broken out to separate tests? We look to have achieved the objective of the test here (redirected to licence summary page and then contact summary page)

File: packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js [hunk 8]
```diff
@@ -220,4 +238,70 @@ describe('The licence summary page', () => {
...
+    const data2 = await injectWithCookies('GET', CONTROLLER.uri)
+    expect(data2.statusCode).toBe(302)
+    expect(data2.headers.location).toBe(NAME.uri)
+  })
+
+  it('date of birth amendment (no licence required) causes a redirect to the no licence required page', async () => {
+    await postRedirectGet(LICENCE_TO_START.uri, { 'licence-to-start': 'after-payment' })
+    const data = await postRedirectGet(DATE_OF_BIRTH.uri, dobHelper(dobJuniorTomorrow))
+    expect(data.statusCode).toBe(302)
+    expect(data.headers.location).toBe(NO_LICENCE_REQUIRED.uri)
```
Comment (jaucourt) [inline packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js:267]: After here, we look to be going beyond the scope of the test; we've checked that we're on the no licence required page

File: packages/gafl-webapp-service/src/pages/summary/contact-summary/__tests__/contact-summary.spec.js (hunk 1 above)
Comment (graham-willis-druid) [reply packages/gafl-webapp-service/src/pages/summary/contact-summary/__tests__/contact-summary.spec.js:84]: done

File: packages/gafl-webapp-service/src/pages/summary/contact-summary/__tests__/contact-summary.spec.js (hunk 2 above)
Comment (graham-willis-druid) [reply packages/gafl-webapp-service/src/pages/summary/contact-summary/__tests__/contact-summary.spec.js:100]: done - data is a bad name yes? Maybe should refactor the entire thing to responseN.

File: packages/gafl-webapp-service/src/pages/summary/contact-summary/__tests__/contact-summary.spec.js (hunk 3 above)
Comment (graham-willis-druid) [reply packages/gafl-webapp-service/src/pages/summary/contact-summary/__tests__/contact-summary.spec.js:148]: I moved a lot of the tests regards dob from the contact to the licence summary. (I forgot to move them before - originally the dob was on the licence summary)

File: packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js (hunk 4 above)
Comment (graham-willis-druid) [reply packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js:123]: done

File: packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js (hunk 5 above)
Comment (graham-willis-druid) [reply packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js:144]: Simplified

File: packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js (hunk 6 above)
Comment (graham-willis-druid) [reply packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js:299]: I am not sure I understand. The juniors cannot get a letter. This is stated in the test description: 'date of birth amendment (junior) causes a method of contact of letter to be set no none'

File: packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js (hunk 7 above)
Comment (graham-willis-druid) [reply packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js:283]: ok - done

File: packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js (hunk 8 above)
Comment (graham-willis-druid) [reply packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js:267]: I guess I got unit test fatigue

File: packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js (hunk 6 above)
Comment (jaucourt) [reply packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js:301]: I just meant that, from the specification of the test, we really just need to test that the preferredMethodOfConfirmation and preferredMethodOfReminder are both 'none' , checking that they're 'letter' as a default is beyond the scope of this test. If you want to test it, it should be in a separate test, I think

Comment (graham-willis-druid) [reply packages/gafl-webapp-service/src/pages/summary/licence-summary/__tests__/licence-summary.spec.js:301]: ok
